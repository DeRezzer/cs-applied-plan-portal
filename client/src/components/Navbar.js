/** @jsx jsx */

import {useState, useEffect} from "react";
import {css, jsx} from "@emotion/core";
import PropTypes from "prop-types";
import {Link} from "react-router-dom";
import {logout} from "../utils/authService";
import {withRouter} from "react-router-dom";
import {getToken} from "../utils/authService";

function Navbar(props) {

  const [notifications, setNotifications] = useState([]);

  const style = css`

    display: flex;
    position: absolute;
    width: 100%;
    height: 35px;
    background-color: orange;

    #navbar-search {
      width: 35%;
    }

    .osu-logo {
      vertical-align: middle;
    }

    .right-container {
      margin-left: auto;
    }

    .dropdown {
      display: inline-block;
    }

    .dropdown .dropbtn, .logout {
      height: 35px;
    }

    .dropdown-content {
      display: none;
      position: absolute;
      background-color: #f9f9f9;
      min-width: 160px;
      box-shadow: 0px 8px 16px 0px rgba(0,0,0,0.2);
      z-index: 1;
    }

    .dropdown-content a {
      float: none;
      color: black;
      padding: 12px 16px;
      text-decoration: none;
      display: block;
      text-align: left;
    }

    .dropdown-content a:hover {
      background-color: #ddd;
    }

    .dropdown:hover .dropdown-content {
      display: block;
    }

    .badge {
      position: absolute;
      top: 5px;
      right: 10px;
      padding: 5px 10px;
      border-radius: 50%;
      background-color: red;
      color: white;
    }
  `;

  useEffect(() => {
    async function fetchData() {

      try {

        const token = getToken();
        const server = `${process.env.REACT_APP_API_HOST}:${process.env.REACT_APP_API_PORT}`;
        const url = `http://${server}/notification` +
          `?accessToken=${token}`;
        let obj = [];

        // get notifications data
        const response = await fetch(url);
        if (response.ok) {
          // get data from the response
          obj = await response.json();
          setNotifications(obj.notifications);
        }

      } catch (err) {
        // log server error, if it happens while fetching notifications
        console.log("An internal server error occurred. Please try again later.");
      }

    }

    fetchData();

  }, []);

  function logoutUser() {
    logout();
    props.history.push("/login");
  }

  return (
    <div className="navbar-parent" css={style}>
      <Link to={"/"}>
        <p className="osu-logo">Oregon State University</p>
      </Link>
      <div className="right-container">
        <div className="dropdown">
          <button className="dropbtn">Notifications
            <i className="fa fa-caret-down"></i>
          </button>
          {notifications.length ?
            <span className="badge">{notifications.length}</span> : null }
          <div className="dropdown-content">
            {notifications.map((item) => (
              <Link key={item.notificationId} to={`/viewPlan/${item.planId}`}>
                {item.text}
              </Link>
            ))}
          </div>
        </div>
        {props.showSearch ? <input id="navbar-search" className="form-control mr-sm-2" type="text" placeholder={props.searchContent} name="search"/> : null}
        <button className="logout" onClick={() => logoutUser()}>Log out</button>
      </div>
    </div>
  );

}
export default withRouter(Navbar);

Navbar.propTypes = {
  history: PropTypes.object,
  showSearch: PropTypes.bool,
  searchContent: PropTypes.string
};