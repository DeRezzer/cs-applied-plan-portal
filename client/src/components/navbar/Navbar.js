/** @jsx jsx */

import {css, jsx} from "@emotion/core";
import {Link} from "react-router-dom";
import {withRouter} from "react-router-dom";
import Notifications from "./Notifications";
import History from "./History";
import Logout from "./Logout";
import {getProfile} from "../../utils/authService";
import {useEffect, useState} from "react";

function Navbar() {

  // role and function to set role, default to 0 (Student)
  const [role, setRole] = useState(0);

  useEffect(() => {
    async function checkRole() {
      const profile = await getProfile();
      if (!profile.role) {
        setRole(0);
      } else {
        setRole(profile.role);
      }
    }
    checkRole();
  }, []);

  const style = css`

    & {
      display: flex;
      position: absolute;
      width: 100%;
      height: 50px;
      background-color: var(--color-orange-500);
      align-items: center;
      padding: 0rem 1rem;
      grid-area: header;
      position: fixed;
      top: 0;
      left: 0;
      right: 0;
      z-index: 9;
    }
    
    & a:first-child:hover {
      text-decoration: none;
    }
    
    & a:first-child {
      text-decoration-color: transparent !important;
    }

    .osu-logo {
      vertical-align: middle;
      font-size: large;
      font-weight: 600;
      margin: 0;
      margin-right: 1rem;
      color: white;
    }

    .right-container {
      margin-left: auto;
    }

    .logout {
      height: 35px;
    }

    @media print {
      & {
        display: none;
      }
    }

  `;

  return (
    <div className="navbar-parent" css={style}>
      <Link to={"/"}>
        <p className="osu-logo">OSU CS Applied Plan Portal</p>
      </Link>
      <div className="right-container">
        {role ? <History /> : null}
        <Notifications />
        <Logout />
      </div>
    </div>
  );

}
export default withRouter(Navbar);
