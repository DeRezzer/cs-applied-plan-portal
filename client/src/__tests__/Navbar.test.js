/* eslint-disable no-undef */
import React from "react";
import {MemoryRouter} from "react-router-dom";
import renderer from "react-test-renderer";

import Navbar from "../components/navbar/Navbar";

describe("Navbar", () => {

  it("renders successfully", () => {
    const wrapper = renderer.create(
      <MemoryRouter>
        <Navbar currentPlan={0} />
      </MemoryRouter>).toJSON();
    expect(wrapper).toMatchSnapshot();
  });
});