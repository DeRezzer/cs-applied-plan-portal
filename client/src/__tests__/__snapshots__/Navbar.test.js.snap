// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navbar renders successfully 1`] = `
<div
  className="navbar-parent css-12g91iz-Navbar"
  id="navbar"
>
  <a
    href="/"
    onClick={[Function]}
  >
    <p
      className="osu-logo"
    >
      OSU CS Applied Plan Portal
    </p>
  </a>
  <div
    className="right-container"
  >
    <div
      className="notification-dropdown css-3i4jcj-Notifications"
    >
      <button
        className="drop-button-notification"
        data-count={0}
      >
        Notifications
        <span
          className="badge"
        />
        <i
          className="fa fa-caret-down"
        />
      </button>
      <div
        className="dropdown-content"
      >
        <a
          href="/"
          onClick={[Function]}
        >
          <p>
            No new notifications
          </p>
        </a>
      </div>
    </div>
    <button
      className="logout-button css-1k1xu2y-Logout"
      onClick={[Function]}
    >
      Log Out
    </button>
  </div>
</div>
`;
